### PR TITLE
Creating MBP version of OSC

### PR DESCRIPTION
**Pull request by yangwill**
Now ready for review!

Implemented MBP version of the operational space controller.

Overall it is identical to the RBT version. The main differences are the different names for the same MBP and RBT method and the fact that MBP uses Context for computing kinematic methods instead of RBT's KinematicsCache.

Although not yet implemented, some optimizations that might be useful include storing the plant contexts instead of recreating them for each QP, and looking into different QP solvers (perhaps using multiple solvers in parallel and returning the first solution?)

### CODE REVIEW

**Comment by mposa** [comment]
Is this awaiting review? I think @yminchen should provide the code review, and I'll just give it a brief look when it's ready.

**Comment by yangwill** [comment]
Once the drake-update branch is merged with master, this should be ready for review.

**Comment by yangwill** [comment]
@yminchen Ready for review.

**Comment by yminchen** [comment]
@yangwill Could you update your PR post (the first comment box), and also mention what the changes are aside from the RBT to MBP change? It will help me to review the PR faster and also make it clearer for whoever read this PR in the future. Thanks!

File: systems/controllers/osc/osc_tracking_data_mbp.h [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+  Eigen::VectorXd ddy_command_sol_;
+
+  // `state_` is the finite state machine state when the tracking is enabled
+  // If `state_` is empty, then the tracking is always on.
+  std::vector<int> state_;
+
+  /// OSC calculates feedback positions/velocities from `tree_w_spr_`,
+  /// but in the optimization it uses `tree_wo_spr_`. The reason of using
+  /// RigidBodyTree without spring is that the OSC cannot track desired
+  /// acceleration instantaneously when springs exist. (relative degrees of 4)
```
Comment (yminchen) [inline systems/controllers/osc/osc_tracking_data_mbp.h:135]: Should update the documentation (both Osc and OscTrackingData)

File: systems/controllers/osc/operational_space_control_mbp.cc [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+  std::unique_ptr<drake::systems::Context<double>> context_wo_spr =
+      plant_wo_spr_.CreateDefaultContext();
+  plant_w_spr_.SetPositionsAndVelocities(context_w_spr.get(), x_w_spr);
+  plant_wo_spr_.SetPositionsAndVelocities(context_wo_spr.get(), x_wo_spr);
+
+  // Get M, f_cg, B matrices of the manipulator equation
+  MatrixXd B = plant_wo_spr_.MakeActuationMatrix();
+  MatrixXd M(n_v_, n_v_);
+  plant_wo_spr_.CalcMassMatrixViaInverseDynamics(*context_wo_spr, &M);
+  const RigidBodyTree<double>::BodyToWrenchMap no_external_wrenches;
```
Comment (yminchen) [inline systems/controllers/osc/operational_space_control_mbp.cc:397]: This should be removed

File: systems/controllers/osc/osc_tracking_data_mbp.h [hunk 3]
```diff
@@ -0,0 +1,377 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <Eigen/Dense>
+#include <drake/multibody/plant/multibody_plant.h>
+
+#include "drake/common/trajectories/exponential_plus_piecewise_polynomial.h"
+#include "drake/common/trajectories/piecewise_polynomial.h"
+#include "drake/multibody/rigid_body_tree.h"
```
Comment (yminchen) [inline systems/controllers/osc/osc_tracking_data_mbp.h:10]: The RBT dependency should be removed

File: systems/controllers/osc/operational_space_control_mbp.cc [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+VectorXd OperationalSpaceControlMBP::SolveQp(
+    const VectorXd& x_w_spr, const VectorXd& x_wo_spr,
+    const drake::systems::Context<double>& context, double t, int fsm_state,
+    double time_since_last_state_switch) const {
+  vector<bool> active_contact_flags = CalcActiveContactIndices(fsm_state);
+
+  std::unique_ptr<drake::systems::Context<double>> context_w_spr =
+      plant_w_spr_.CreateDefaultContext();
+  std::unique_ptr<drake::systems::Context<double>> context_wo_spr =
+      plant_wo_spr_.CreateDefaultContext();
```
Comment (yminchen) [inline systems/controllers/osc/operational_space_control_mbp.cc:389]: Making `context_w_spr` and `context_wo_spr` members of OSC can probably boost the speed a little bit (we don't need to allocate the memory for the context every time SolveQp is called)

File: systems/controllers/osc/osc_tracking_data_mbp.h [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+  Eigen::VectorXd GetDdyCommandSol() { return ddy_command_sol_; }
+
+  // Update() updates the caches. It does the following things in order:
+  //  - update track_at_current_state_
+  //  - update desired output
+  //  - update feedback output (Calling virtual methods)
+  //  - update command output (desired output with pd control)
+  // Inputs/Arguments:
+  //  - `x_w_spr`, state of the robot (with spring)
+  //  - `cache_w_spr`, plant context of the robot (without spring)
```
Comment (yminchen) [inline systems/controllers/osc/osc_tracking_data_mbp.h:69]: typo here. cache_w_spr -> context_w_spr

File: systems/controllers/osc/osc_tracking_data_mbp.h [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  int n_r_;
+
+  // PD control gains
+  Eigen::MatrixXd K_p_;
+  Eigen::MatrixXd K_d_;
+
+  // Cost weights
+  Eigen::MatrixXd W_;
+
+  // cache
```
Comment (yminchen) [inline systems/controllers/osc/osc_tracking_data_mbp.h:176]: This comment should probably be removed (doesn't fit here)

File: systems/controllers/osc/osc_tracking_data_mbp.h (hunk 1 above)
Comment (yangwill) [reply systems/controllers/osc/osc_tracking_data_mbp.h:128]: Done.

File: systems/controllers/osc/operational_space_control_mbp.cc (hunk 2 above)
Comment (yangwill) [reply systems/controllers/osc/operational_space_control_mbp.cc]: Done

File: systems/controllers/osc/operational_space_control_mbp.cc (hunk 4 above)
Comment (yangwill) [reply systems/controllers/osc/operational_space_control_mbp.cc]: Done.

File: systems/controllers/osc/osc_tracking_data_mbp.h (hunk 6 above)
Comment (yangwill) [reply systems/controllers/osc/osc_tracking_data_mbp.h]: Done.

File: systems/controllers/osc/osc_tracking_data_mbp.h (hunk 5 above)
Comment (yangwill) [reply systems/controllers/osc/osc_tracking_data_mbp.h]: Done.

**Comment by yangwill** [comment]
Ready for final review

**Comment by mposa** [comment]
I'm a little concerned about the file/class names here. Should we move the RBP version to /attic/ and then this won't need the _mbp prefix?

**Pull request merged**